File: TaskApp/src/main/java/com/example/springboottaskmanager/controller/DepartmentController.java
package com.example.springboottaskmanager.controller;

import com.example.springboottaskmanager.aspect.Auditable;
import com.example.springboottaskmanager.exception.body.NotFoundException;
import com.example.springboottaskmanager.exception.body.UnknownException;
import com.example.springboottaskmanager.service.departmentservices.DepartmentServiceGet;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.ResourceAccessException;

/**
 * Контроллер, передающий GET запросы другому сервису,
 * работающему с департаментами пользователей.
 */
@RestController
@Validated
@RequiredArgsConstructor
@RequestMapping("/taskApp/departments")
public class DepartmentController {
    /**
     * Переменная окружения, определяющая способ обращения к наружнему сервису.
     */
    @Value("${department_client}")
    private String departmentClient;
    /**
     * Сервис департаментов.
     */
    private final DepartmentServiceGet departmentServiceGet;

    /**
     * Метод вызывает другой метод класса {@link DepartmentServiceGet}
     * для получения департамента пользователя.
     * @param userId Идентификатор пользователя.
     * @return Департамент пользователя.
     */
    @Auditable
    @GetMapping("/get/{userId}")
    public String getUserDepartment(@PathVariable final String userId) {
        try {
            return switch (departmentClient) {
                case "FeignClient" -> departmentServiceGet
                        .getUserDepartmentFeign(userId);
                case "WebClient" -> departmentServiceGet
                        .getUserDepartmentWebClient(userId);
                case "RestTemplate" -> departmentServiceGet
                        .getUserDepartmentRestTemplate(userId);
                default -> "Переменная department_client может быть:"
                       + "RestTemplate, FeignClient или WebClient";
            };
        } catch (HttpClientErrorException httpClientErrorException) {
            throw new NotFoundException(httpClientErrorException.getMessage());
        } catch (ResourceAccessException resourceAccessException) {
            throw new UnknownException("Сервис подразделений недоступен");
        }
    }
}

File: DepartmentApp/src/main/kotlin/com/example/department_app/controller/DepartmentController.kt
package com.example.department_app.controller

import com.example.department_app.exception.UnknownException
import com.example.department_app.service.DepartmentService
import org.apache.commons.lang3.math.NumberUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/departments")
class DepartmentController{
    @Autowired
    private lateinit var departmentService: DepartmentService

    var errorMessage: String = "Requested id is not valid"

    @PostMapping("/post/{departmentId}/{userId}")
    fun postUserDepartment(@PathVariable departmentId: String, @PathVariable userId: String): String {
        if(NumberUtils.isCreatable(userId) && NumberUtils.isCreatable(departmentId))
            return departmentService.postUserDepartment(userId.toLong(), departmentId.toLong())
        else
            throw UnknownException(errorMessage)
    }

    @GetMapping("/get/{userId}")
    fun getDepartment(@PathVariable userId: String): String {
        if(NumberUtils.isCreatable(userId))
            return departmentService.getDepartment(userId.toLong())
        else
            throw UnknownException(errorMessage)
    }

    @DeleteMapping("/delete/{userId}")
    fun deleteUser(@PathVariable userId: String): String {
        if(NumberUtils.isCreatable(userId))
            return departmentService.deleteUser(userId.toLong())
        else
            throw UnknownException(errorMessage)
    }
}